pkg/pgmodel/invalidation.readme.md

```diff
@@ -0,0 +1,101 @@
+# Epoch-based series cache invalidation
+
+This document describes the mechanism used to maintain the series cache in
+Promscale.
+
+## Background
+
+Each time series in Prometheus is uniquely identified by the set of labels
+associated with it. In Promscale, we create an entry in the
+`_prom_catalog.series` table for each series. Thus, each series has a unique
+identifier in the database.
+
+The Promscale series cache contains a lookup from series labels to series id.
+It is heavily used while ingesting data, and drastically increases ingest
+throughput (from ~10k samples/sec to ~400k samples/sec).
+
+Rows in the `_prom_catalog.series` table can be removed. This happens when all
+data belonging to a series has been removed (e.g. through the application of
+data retention policies, or manual deletion). The process of deleting a series
+takes place in the database, a process over which the Promscale connector has
+no control. It is possible for a series to be recreated once deleted. When this
+happens, the series id is not the same as before.
+
+To improve insertion performance, there is no foreign key between the data
+table, and the `_prom_catalog.series` table. Inserting samples in the data
+table with the incorrect series id will result in inconsistent data in the
+database.
+
+This implies that we require a mechanism to ensure that entries in the series
+cache are invalidated when the underlying row is removed from the database.
+This is the cache invalidation mechanism.
+
+## Soft-deletion of series
+
+In order to allow time for the synchronisation of deleted series rows between
+the database and the Promscale connector, we soft-delete a series in the
+database. Once the series row has been soft-deleted for long enough, it is
+removed. "Long enough" is configured by the `epoch_duration` parameter, which
+is 12 hours by default.
+
+In the database, we track two unix timestamps in the `_prom_catalog.ids_epoch`
+table: `current_epoch`, and `delete_epoch`.
+`current_epoch` contains the current timestamp (in seconds since the unix epoch)
+at the most recent update.
+`delete_epoch` contains a timestamp marking which rows were removed (this will
+become clear shortly).
+Note: using `current_epoch` as the fixed notion of when "now" is gets around
+issues of time synchronisation between different systems.
+
+The soft-deletion works as follows: When a row is to be removed, we set the
+`delete_epoch` column of `_prom_catalog.series` to the value currently
+contained in the `current_epoch` column of the `_prom_catalog.ids_epoch` table.
+This indicates: it was decided at `current_epoch` to delete this row at some
+point in the future.
+
+The row is removed when the following condition is met:
+`prom_catalog.series.delete_epoch` >= `current_epoch` - `epoch_duration`
+When the row is removed, `_prom_catalog.ids_epoch.delete_epoch` is set to
```

Comment:
These two sentences aren't wrong and convey the system's intent, but I believe they misrepresent the causality of events taking place during an invalidation. 

We indeed for a row to be removed when the following condition is met: `prom_catalog.series.delete_epoch` >= `current_epoch` - `epoch_duration`. To make the system's state more explicit and allow for reconfigurations we track `current_epoch` - `epoch_duration` as `_prom_catalog.ids_epoch.delete_epoch`. Therefore a row is considered safe to delete when `prom_catalog.series.delete_epoch` >= `_prom_catalog.ids_epoch.delete_epoch`

Reply:
Yes, you're right - thank you!